File: clients/rs/src/features/wrap_type_json.rs
use polywrap_client::{
    client::PolywrapClient, core::uri::Uri, msgpack::wrappers::polywrap_json::JSONString,
};
use polywrap_client_default_config::SystemClientConfig;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::error::Error;

#[derive(Serialize, Deserialize)]
#[serde(untagged)]
enum Args {
    Parse(ParseArgs),
    Stringify(StringifyArgs),
}

#[derive(Serialize, Deserialize)]
struct ParseArgs {
    value: JSONString,
}

#[derive(Serialize, Deserialize)]
struct StringifyArgs {
    values: Vec<JSONString>,
}

#[derive(Serialize, Deserialize)]
struct InputObj {
    method: String,
    args: Args,
}

pub fn run_test_case(input: &Value) -> Result<(), Box<dyn Error>> {
  let input_obj: InputObj = serde_json::from_value(input.clone())?;
    let method = input_obj.method;
    let args = match input_obj.args {
        Args::Parse(a) => polywrap_client::msgpack::to_vec(&a)?,
        Args::Stringify(a) => polywrap_client::msgpack::to_vec(&a)?,
    };

    let binding = std::env::current_dir()?.join("../../wraps");
    let root = binding.to_str().unwrap();
    let uri: Uri = format!("fs/{root}/json-type/implementations/as")
        .try_into()
        .unwrap();

    let client: PolywrapClient = PolywrapClient::new(SystemClientConfig::default().into());

    println!("Invoking {method}");

    let result = client.invoke::<String>(&uri, &method, Some(&args), None, None);

    match result {
        Ok(result) => {
            println!("Result: {result:?}");
            println!("Success!");
        }
        Err(err) => panic!("{err}"),
    }

    Ok(())
}

File: clients/py/src/features/wrap_type_json.py
from pathlib import Path
from typing import Any, List, TypedDict, Union

from polywrap import (
    FsUriResolver,
    PolywrapClient,
    PolywrapClientConfigBuilder,
    SimpleFileReader,
    Uri,
)
from pydantic import BaseModel


class ArgsParse(TypedDict):
    value: str


class ArgsStringify(TypedDict):
    values: List[str]


class TestCaseInput(BaseModel):
    method: str
    args: Union[ArgsParse, ArgsStringify]


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.parse_obj(input)

    root = Path(__file__).parent.parent.parent.parent.parent / "wraps"
    uri = Uri.from_str(f"fs/{root}/json-type/implementations/as")

    config = (
        PolywrapClientConfigBuilder()
        .add_resolver(FsUriResolver(SimpleFileReader()))
        .build()
    )

    client = PolywrapClient(config)

    print(f"Invoking {input_obj.method}")

    response = client.invoke(
        uri=uri,
        method=input_obj.method,
        args=input_obj.args,
    )

    if not response:
        raise ValueError(f"Error: {response}")

    print("Result:", response)
    print("Success!")
